### PR TITLE
[SLP]Try detect strided loads, if any pointer op require extraction.

### PR DESCRIPTION
**Pull request by alexey-bataev**
If any pointer operand of the non-cosencutive loads is an instructions
with the user, which is not part of the current graph, and, thus,
requires emission of the extractelement instruction, better to try to
detect if the load sequence can be repsented as strided load and
extractelement instructions for pointers are not required.

### CODE REVIEW

**Comment by alexey-bataev** [comment]
Ping!

File: llvm/test/Transforms/SLPVectorizer/RISCV/complex-loads.ll [hunk 1]
```diff
@@ -1,213 +1,225 @@
 ; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 3
-; RUN: opt -S -mtriple riscv64-unknown-linux-gnu < %s --passes=slp-vectorizer -mattr=+v -slp-threshold=-20 | FileCheck %s
+; RUN: opt -S -mtriple riscv64-unknown-linux-gnu < %s --passes=slp-vectorizer -mattr=+v -slp-threshold=-15 | FileCheck %s
```
Comment (RKSimon) [inline llvm/test/Transforms/SLPVectorizer/RISCV/complex-loads.ll:2]: Worth keeping the 20 threshold as well?

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -4617,12 +4617,21 @@ BoUpSLP::LoadsState BoUpSLP::canVectorizeLoads(
...
+          (((Sz > MinProfitableStridedLoads ||
+             (static_cast<unsigned>(std::abs(*Diff)) <=
+                  MaxProfitableLoadStride * Sz &&
+              isPowerOf2_32(std::abs(*Diff)))) &&
+            static_cast<unsigned>(std::abs(*Diff)) > Sz) ||
+           *Diff == -(static_cast<int>(Sz) - 1) ||
+           any_of(PointerOps, [&](Value *V) {
+             return isa<Instruction>(V) && any_of(V->users(), [&](User *U) {
+                      return !getTreeEntry(U) && !MustGather.contains(U);
+                    });
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:4633]: (observation not request) - its very tricky to check the logic in this now, but I'm not sure whether its worth pulling this out into a helper.

**Review by RKSimon — APPROVED** [review]
LGTM

**Pull request merged**